astroid/helpers.py

```diff
@@ -244,6 +246,23 @@ def object_len(node, context=None):
     from astroid.objects import FrozenSet
 
     inferred_node = safe_infer(node, context=context)
+
+    # prevent self referential length calls from causing a recursion error
+    # see https://github.com/PyCQA/astroid/issues/777
+    node_frame = node.frame()
+    if (
+        isinstance(node_frame, scoped_nodes.FunctionDef)
+        and node_frame.name == "__len__"
+        and inferred_node._proxied == node_frame.parent
+    ):
+        message = (
+            "Self referential __len__ function will "
+            "cause a RecursionError on line {} of {}".format(
+                node.lineno, node.root().file
+            )
+        )
+        raise exceptions.AstroidRecursionError(message)
```

Comment:
Rather than adding a new exception, let's stop with `InferenceError` like the rest of the function is doing.